apps/test/unit/lib/levelbuilder/rubrics/RubricsContainerTest.js

```diff
@@ -16,13 +17,86 @@ describe('RubricsContainerTest', () => {
     lessonNumber: 0,
   };
 
+  const rubricInfo = {
+    learningGoals: [
+      {
+        key: 'ui-1',
+        id: 'ui-1',
+        learningGoal: '',
+        aiEnabled: false,
+        position: 1,
+        learningGoalEvidenceLevelsAttributes: [
+          {
+            teacherDescription: '',
+            understanding: RubricUnderstandingLevels.NONE,
+            aiPrompt: '',
+          },
+          {
+            teacherDescription: '',
+            understanding: RubricUnderstandingLevels.LIMITED,
+            aiPrompt: '',
+          },
+          {
+            teacherDescription: '',
+            understanding: RubricUnderstandingLevels.CONVINCING,
+            aiPrompt: '',
+          },
+          {
+            teacherDescription: '',
+            understanding: RubricUnderstandingLevels.EXTENSIVE,
+            aiPrompt: '',
+          },
+        ],
+      },
+      {
+        key: 'ui-2',
+        id: 'ui-2',
+        learningGoal: '',
+        aiEnabled: false,
+        position: 2,
+        learningGoalEvidenceLevelsAttributes: [
+          {
+            teacherDescription: '',
+            understanding: RubricUnderstandingLevels.NONE,
+            aiPrompt: '',
+          },
+          {
+            teacherDescription: '',
+            understanding: RubricUnderstandingLevels.LIMITED,
+            aiPrompt: '',
+          },
+          {
+            teacherDescription: '',
+            understanding: RubricUnderstandingLevels.CONVINCING,
+            aiPrompt: '',
+          },
+          {
+            teacherDescription: '',
+            understanding: RubricUnderstandingLevels.EXTENSIVE,
+            aiPrompt: '',
+          },
+        ],
+      },
+    ],
+  };
+
   it('renders the components on the page correctly for a new rubric', () => {
     const wrapper = mount(<RubricsContainer {...defaultProps} />);
+    expect(wrapper.find('Heading1').text()).to.equal('Create your rubric');
     expect(wrapper.find('select#rubric_level_id option')).to.have.length(3);
     expect(wrapper.find(LearningGoalItem)).to.have.length(1);
     expect(wrapper.find('Button[text="Save your rubric"]')).to.have.length(1);
   });
 
+  it('renders "the components on the page correctly for an exisiting rubric"', () => {
+    const props = {...defaultProps, rubric: rubricInfo};
+    const wrapper = mount(<RubricsContainer {...props} />);
+    expect(wrapper.find('Heading1').text()).to.equal('Modify your rubric');
+    expect(wrapper.find('select#rubric_level_id option')).to.have.length(3);
+    expect(wrapper.find(LearningGoalItem)).to.have.length(2);
+    expect(wrapper.find('Button[text="Save your rubric"]')).to.have.length(1);
+  });
+
```

Comment:
These tests are all pretty aligned to expectations. I generally try to think of extremes... things I might be missing in my manual testing or design. Like, what happens when it has no levels? Is there anything we should test, there?

Also, I'm looking for a test to cover every interaction and state. We do not have a test for the `handleDropdownChange` action nor do we test the enabling/disabling of the save button. I think those would be good candidates for tests. Are there any others?

Reply:
Looking at your candidates:

1**. What happens when it has no levels:** We do have a [ticket](https://codedotorg.atlassian.net/jira/software/projects/AITT/boards/70?selectedIssue=AITT-170) for if there are no levels, a user should never get to this page.  That being said, in a different PR (which will have a test for this), I am planning on making some UI changes to let the user know they need to have a submittable level to be able to have a rubric.
2. **We do not have a test for the `handleDropdownChange` action** - Correct, this also corresponds to a state change.  I created a test for this.  
3. **Test the enabling/disabling of the save button.**  This test was tricky.  I added a test for this (and the other UI changes that happen when hitting the "save" button).  I did some research into the `async` `await` pattern and found a solution.  We use something similar in two other places in our code base.  From what I have seen so far in my time on the engineering side of things, it seems like this case would be covered in an integration test of some type.  At this point we don't have any integration tests for this work flow.  [I created a ticket for it.](https://codedotorg.atlassian.net/jira/software/projects/AITT/boards/70?selectedIssue=AITT-177)
4.  **I'm looking for a test to cover every interaction and state** - I think this PR covers all of the rendered components that are interact-able (The button and the dropdown) and also covers all components that render conditionally (the notified text and the button being disabled).  On the state side it gets a bit tricky.  We certainly cover when the state of `saveNotificationText` or the `selectedLevelForAssessment` is changed.  However, the other state (`learningGoalList`) is largely there so the information can be passed down to other child components.  So the state of this will not change much in this component alone.  Again, for this reason it seems like some integration tests would be helpful.

Reply:
I don't think I would use integration tests to test this kind of state. Do we have examples of using integration testing this way? I would probably still use traditional unit testing to test state logic that is passed down to child components. I wrote a looooooong comment that went through and justified this and what I would do, haha, but then I erased it because it was very involved. Basically, I would strive to unit test each component and mock out the state handlers being passed down. This means unit testing the state logic at its source (the RubricsContainer) and testing that it is passed down. Then unit testing the other components such that they properly call them. I don't value integration testing these together (on top of also unit testing them) all that much.

Reply:
Happy to share my short novel lol. I'm reflecting and think I might be a bit over-opinionated and surely would be happy with the best solution given the time and value. But I'm happy to pair on this because I think it's tricky for either of us.